src/ksym.rs

```diff
@@ -91,29 +91,26 @@ impl KSymResolver {
         }
     }
 
-    pub fn find_address_ksym(&self, addr: u64) -> Option<&Ksym> {
+    pub fn find_addresses_ksym(&self, addr: u64) -> Vec<&Ksym> {
         let mut l = 0;
         let mut r = self.syms.len();
 
-        if !self.syms.is_empty() && self.syms[0].addr > addr {
-            return None;
-        }
-
-        while l < (r - 1) {
-            let v = (l + r) / 2;
-            let sym = &self.syms[v];
+        while l < r {
+            let m = (l + r) / 2;
+            let sym = &self.syms[m];
 
-            if sym.addr == addr {
-                return Some(sym);
-            }
             if addr < sym.addr {
-                r = v;
+                r = m;
             } else {
-                l = v;
+                l = m + 1;
             }
         }
-
-        Some(&self.syms[l])
+        // `self.syms[l - 1].addr <= addr < self.syms[l].addr`.
```

Comment:
The new implementation is basically the first part of the previous revision, but remove the check of the exact match.  It is almost the same as what @Andrii provided; however, I changed the condition of moving left and right a little bit.

Here it moves left (`l`) when `addr >= sym.addr` instead of `addr > sym.addr`.
Moving left when `addr > sym.addr` would make 
 - `l` points at **the first one of the matched symbols** for the **exactly matched** cases, while 
 - `l` points to **the position after the matched symbols** for the cases that the address of the matched symbols is **smaller**.

Reply:
So I added the following test:
```rust
    #[test]
    fn foobar() {
        let mut resolver = KSymResolver::new();
        resolver.syms = vec![
          Ksym { addr: 0x123, name: "1".to_string(), c_name: RefCell::new(None), },
          Ksym { addr: 0x123, name: "1.5".to_string(), c_name: RefCell::new(None), },
          Ksym { addr: 0x1234, name: "2".to_string(), c_name: RefCell::new(None), },
          Ksym { addr: 0x12345, name: "3".to_string(), c_name: RefCell::new(None), },
        ];
        resolver.syms.sort_by(|a, b| a.addr.cmp(&b.addr));

        assert_eq!(resolver.find_addresses_ksym(1), Vec::<&Ksym>::new());
        assert_eq!(resolver.find_addresses_ksym(0x123), vec![
          &Ksym { addr: 0x123, name: "1.5".to_string(), c_name: RefCell::new(None), },
          &Ksym { addr: 0x123, name: "1".to_string(), c_name: RefCell::new(None), },
        ]);

        assert_eq!(resolver.find_addresses_ksym(0x124), Vec::<&Ksym>::new()); // fails
        assert_eq!(resolver.find_addresses_ksym(0x1234568), Vec::<&Ksym>::new()); // fails
    }
```

Is it expected that a search for 0x124 returns symbols at 0x123? What about a search for 0x1234568 returning the symbol at 0x12345? It's not clear to me why that would be desired behavior, but I could be missing something. The documentation seems to suggest the following:
```
    /// Find the names and the start addresses of a symbol found for
    /// the given address.
```
Perhaps I am misreading it, but I am not reading anything about the intention to return potentially completely unrelated symbols.

Assuming the binary search bits check out, I don't really understand the `take_while` condition. Wouldn't something like the following be more appropriate (which incidentally makes the above tests pass, which would certainly be more closely resembling my expectations of the API)?
```patch
--- src/ksym.rs
+++ src/ksym.rs
@@ -109,7 +110,7 @@ impl KSymResolver {
         self.syms[0..l]
             .iter()
             .rev()
-            .take_while(|sym| sym.addr == self.syms[l - 1].addr)
+            .take_while(|sym| sym.addr >= addr)
             .collect()
     }
 }
```

Irrespective of the answer to this last question, I urge you to add a proper set of unit tests with well known inputs so that we can at least have some faith in this code.

Reply:
I am not sure how you test it.  Since I don't implement or derive `Debug` trait, I rewrote the test case you have to test.

```
     #[test]      
    fn foobar() {
        let mut resolver = KSymResolver::new();
        resolver.syms = vec![                                                        
            Ksym { addr: 0x123, name: "1".to_string(), c_name: RefCell::new(None), },
            Ksym { addr: 0x123, name: "1.5".to_string(), c_name: RefCell::new(None), },
            Ksym { addr: 0x1234, name: "2".to_string(), c_name: RefCell::new(None), },
            Ksym { addr: 0x12345, name: "3".to_string(), c_name: RefCell::new(None), },
        ];                                                                                                                              
        //resolver.syms.sort_by(|a, b| a.addr.cmp(&b.addr));                                                                                                                                                                                                                     
                                                            
        assert!(resolver.find_addresses_ksym(1).is_empty());
        let syms = resolver.find_addresses_ksym(0x123);
        assert_eq!(syms.len(), 2);
        assert_eq!(syms[0].addr, 0x123);
        assert_eq!(syms[0].name, "1.5");
        assert_eq!(syms[1].addr, 0x123);
        assert_eq!(syms[1].name, "1");
                                                       
        let syms = resolver.find_addresses_ksym(0x124);
        assert_eq!(syms.len(), 2);
        assert_eq!(syms[0].addr, 0x123);
        assert_eq!(syms[0].name, "1.5");
        assert_eq!(syms[1].addr, 0x123);
        assert_eq!(syms[1].name, "1");                     
        let syms = resolver.find_addresses_ksym(0x1234568);
        assert_eq!(syms.len(), 1);        
        assert_eq!(syms[0].addr, 0x12345);
        assert_eq!(syms[0].name, "3");
    }
```
And,  it pass.

I have mentioned the format of kallsyms in the earlier discussions.  The kallsyms is a list of (start address, symbol name) pairs.  They (pairs) don't include the end addresses or the sizes of symbols or objects.  So, we are finding two symbols that the given address (`addr`) is falling in between their start addresses and returning the earlier one.

For example, `0x123` & `0x124` in your case is falling in between `resolver.syms[1]` and `resolver.syms[2]`, so `resolver.syms[1]` is returned.  However, to handle multiple symbols at the same address, it returns `resolvers.syms[0]` as well.

For the case `0x12345678`, it is bigger than the last symbol.  So, it deems the last symbol covers this address.

Using `sym.addr >= addr` as the `take_while` condition will return all symbols in `self.syms[0..l]`.

For all `i`, `self.syms[i].addr >= addr` will be true where
 - `i < l` ,
 - `self.syms[l - 1].addr <= addr < self.syms[l].addr`, and
 - `self.syms` is sorted.

Reply:
Gotcha. That's kind of what I later came up with as an explanation as well. Can we *please* document that behavior properly, though? I.e., at the very least we should explain that for the very method in question. But I think the `SymResolver` trait should also mention that the results are reported on a best effort basis and some resolvers may return symbols that are not guaranteed to match an address. Ideally we'd encode that better in the type system, but let's do one step at a time.

Reply:
Overall the implementation is much cleaner and easier to understand than in the first draft, though. That's great!

Reply:
I added the test case to the code.
It tests the cases of addresses that are
 - < first address,
 - == the first address,
 - in between the first and second addresses,
 - == an address in the middle,
 - in between two addresses,
 - == the last address, and
 - &gt;  than the last address.